typed-lambda-calculus/article.md

```diff
@@ -0,0 +1,1016 @@
+Arguably, this insight kickstarted the research on functional programming languages.
+
+In 1969, William Alvin Howard finalized the idea initially proposed by Haskell Curry in 1934, which equated intuitionistic logic systems (logic systems without the law of excluded middle) with typed lambda calculi.
+This kickstarted the research to find new type theories for existing logics and new logics for existing type theories.
+
+In 1972, Per Martin-Löf found a type theory corresponding to predicate logic by introducing dependent types.
+This theory became known as intuitionistic type theory or commonly Martin-Löf type theory.
+
+In 1986, Thierry Coquand and Gérard Huet created the calculus of constructions, a dependent type theory for functions.
+This theory became the basis for Coq and Lean proof assistants.
+
+Since the early 2000s, type theory research has mostly focused on the homotopy type theory (HoTT).
+In HoTT the notion of equivalence is generalized to homotopy -- i.e. the topological notion of similarity.
+HoTT turns out to be powerful enough to discuss the foundations of mathematics again and is used in the univalent foundations project, which is arguably somewhat of a spiritual successor to Hilbert's program.
+
+## Simply typed lambda calculus
+
+Untyped lambda calculus contains some "puns". In the previous article, we noted that the term $fls$ (denoting boolean zero in Church encoding) and the term $c_0$ (denoting natural zero in Church encoding) have the same representation.
+
+Furthermore, untyped lambda calculus allows for constructions like, for example, $plus\; tru\; c_2,$ which don't have a defined semantic meaning.
+
+Various typed lambda calculi introduce some restrictions on the form that terms in different expressions might have, as to avoid this punning and meaningless terms.
+
+### Lambda calculus with atomic boolean values
+
+To facilitate the discussion, let us introduce a couple additional atomic value-terms to our lambda calculus syntax, namely, values $true$ and $false$.
+This, of course, makes our lambda calculus impure.
+
+The reason we need to do this will become apparent later, but for now, suffice it to say that we can't have simply typed pure lambda calculus as the "simple" type system is not expressive enough.
+
+Since we introduced atomic values, we also need to introduce an eliminator (i.e. consumer) for these values, in this case let's consider an $\mathbf{if} \ldots \mathbf{then} \ldots \mathbf{else} \ldots$ operator.
+
+The semantics of this operator should be reasonably obvious.
+We'll only note that there are two meaningful options here: either both branches are computed (corresponding to "eager" evaluation) or only the matching one is (corresponding to "lazy" evaluation).
+In practice, programming languages generally prefer the second option for obvious reasons, but the exact choice doesn't affect the following discussion.
```

Comment:
Not obvious to me. I mean, yeah, to avoid unnecessary computation — but then why stop at boolean operators? Why not make all computations short-circuit? In fact, Haskell does exactly that. There must be a reason why the majority of programming languages opts for lazy evaluation of boolean operators but nothing else. It's probably very subtle and not obvious at all.

And if it doesn't affect the discussion, just don't mention it.